### ISSUE TITLE
Tracking issue for `overlapping_inherent_impls` compatibility lint

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#36889]
**UPDATE:** [We've decided to convert this warning into a hard error](https://github.com/rust-lang/rust/issues/36889#issuecomment-275166478), but are still waiting for a PR to make that a reality. If you're interested in contributing, [there are directions for writing such a PR available here](https://github.com/rust-lang/rust/issues/36889#issuecomment-287145267).

This tracking issue pertains to a bug fix for https://github.com/rust-lang/rust/issues/22889. The problem was that the compiler was accepting two inherent impls that both defined the same method, which creates an obvious ambiguity:

```rust
struct Foo;

impl Foo {
    fn id() {}
}

impl Foo {
    fn id() {}
}
```

The most obvious fix here is to give one method a distinct name, though sometimes there are other possible rewrites (e.g., one could rewrite the impls to use a double dispatch pattern).

#### Current status

- [x] https://github.com/rust-lang/rust/pull/31925 introduces the `overlapping_inherent_impls` lint as warn-by-default
- [x] https://github.com/rust-lang/rust/pull/36894 makes the `overlapping_inherent_impls` lint deny-by-default
- [ ] PR ? makes the `overlapping_inherent_impls` lint a hard error

**Comment by petrochenkov** [comment]
@aturon , could you write a description for this?

**Comment by QuietMisdreavus** [comment]
#38967 could be related to this.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

This warning has been out for a while. I propose we move it to **deny-by-default**.

**Comment by petrochenkov** [comment]
@nikomatsakis 
>I propose we move it to deny-by-default.

There's a catch.
It's already deny-by-default for almost 4 months.

**Comment by nikomatsakis** [comment]
@petrochenkov heh =) Um, d'oh! Then I propose we kill it. I'll start again. :)

**Comment by nikomatsakis** [comment]
@rfcbot fcp cancel

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

This warning has been deny-by-default for a while. I propose we move it to a **hard error**.

**Comment by aturon** [comment]
cc @petrochenkov @rust-lang/compiler -- anyone want to make a quick stabilization PR that we can backport to 1.17 beta?

**Comment by nikomatsakis** [comment]
This seems like a good first bug. I wrote up some directions and posted them to forge for converting compatibility lints into hard errors. The descriptions probably haven't landed yet, but here is the PR:

https://github.com/rust-lang-nursery/rust-forge/pull/48

Marking a E-easy and E-mentor, the directions are in that PR. =)

**Comment by topecongiro** [comment]
Hello, I would like to work on this if no one is interested.

**Comment by nikomatsakis** [comment]
@topecongiro go for it:)

**Comment by nikomatsakis** [comment]
@topecongiro please feel free to ping me on irc (`nmatsakis`) if you encounter any difficulty...

**Comment by topecongiro** [comment]
@nikomatsakis Thank you!
